update ios and android apps for GreenAddress rebranded to Blockstream…

greenaddress opened the pull request:
… Green

Hi, I hope I did all the updates correctly, if not, please let me know and I'll update asap.

I removed the GreenAddress iOS app as it has been replaced by Blockstream Green for iOS.
I also removed GreenBits (for Android) as likewise it has been replaced by Blockstream Green for Android.

We compressed the new screenshot and logo, let me know if anything is missing.

crwatkins commented:
Thanks @greenaddress for the updates and for the new wallets.  We have a listing criterium that specifies
* Wallet was publicly announced and released since at least 3 months

Since this is a brand new codebase and new repository, my recommendation would be that this criterium applies to Blockstream Green and thus would be eligible for listing on 19 June 2019 to allow for adequate community review and usage.  We can certainly start the review process in the meantime.

The [announcement](https://blockstream.com/2019/03/19/the-all-new-blockstream-green-wallet/) states 
> We’re going to be gradually retiring our trusty Greenbits and GreenAddress wallets [...]

@greenaddress Is it reasonable to continue listing them until Blockstream Green replaces them or are they no longer available?  What do you recommend?

One last technical question:  Do the new wallets support both 2of2 and 2of3 modes?

greenaddress commented:
It's all based on libwally like it used to - there is new code and certainly more than other releases but for Android especially we started off from GreenBits - we just moved from just using libwally to share some more code with libgdk and the UX has been redone.

This is the update to the old apps + rebranding. The featureset is the ~ the same and the libraries are mostly the same (at least for signing, we have been using libwally for years now), although we did change transaction construction (but we did this in the past too from server side to client side without this changing the bitcoin.org listing)

I am happy to wait for 3 months but given the above I think it would be a disservice to the users (provided review is completed by then of course)

These new UX wallets support 2of2 and 2of3, for 2of3 you have to create them on the desktop wallet first (as it has always been for GreenBits at least)

Cobra-Bitcoin commented:
It seems like a really significant change given that the old repository was abandoned and this has been worked on for a long time. I would argue that this is closer to a new wallet listing rather than an update of an existing one and it therefore needs time to establish itself.

> I am happy to wait for 3 months but given the above I think it would be a disservice to the users (provided review is completed by then of course)

The real disservice to the users would be to list the wallet without proper review or letting enough time pass to allow for production problems with the wallet to be ironed out. 

Also, personally, as this is a Blockstream wallet, I'm inclined to not trust it too easily without extensive review, since I consider the company a minor nefarious actor in the space, mainly due to the events of the "Hong Kong agreement" and social media manipulation.

greenaddress commented:
@Cobra-Bitcoin as you prefer. You have your reasons.

From my point of view this is slightly strange.

We have done more extensive changes in the past when we moved off from server side transaction construction into client side, when we moved off bitcoinjs and bitcoinj and pycoin signing into libsecp256k1 first and then a wrapper there of (libwally). and now this is just the next update. 

Given the above, I suspect you don't review wallet repo changes across time nor notice any major refactoring nor change of library - as per above - (and you can check the history yourself the old repos are all there to see how much we changed over the years.)

And I am not even suggesting any review or due process is skipped - just surprised a name changed triggered this is probably not very consistent or accurate way of enforcing the rules you set.

greenaddress commented:
3 months have passed since the PR 

@crwatkins did you have a chance to review the wallet/codebase? any feedback?

Would you like me to rebase the PR?

thanks

crwatkins commented:
@greenaddress It's next up; we just finished BLW hours ago.  It should go faster than most because of the legacy.

A rebase would be swell.  Thanks!

greenaddress commented:
@crwatkins rebased, cheers

crwatkins commented:
I have reviewed Blockstream Green based on the current wallet requirements criteria and my evaluation is below. The summary is that the wallet passes on security and overall design and I recommend it for listing.

The scoring in 2c0d05d for the Android wallet is **checkpassvalidationspvp2p** while the scoring for the iOS wallet is **checkpassvalidationservers**.  There are no SPV settings on the iOS wallet.  @greenaddress what does the iOS wallet use for validation?  If it uses similar validation to GreenAddress wallets, then the scoring should be the same **checkfailvalidationcentralized** as GreenAddress. Once this is resolved, I would be glad to recommend this PR for merging.


## Blockstream Green
### iOS version 3.1.1
### Android version 3.1.3
### Review Version 2019063001


The wallet list is based on the personal evaluation of the maintainer(s) and
regular contributors of this site, according to the criteria detailed below.

These requirements are meant to be updated and strengthened over time.
Innovative wallets are exciting and encouraged, so if your wallet has a good
reason for not following some of the rules below, please submit it anyway and
we'll consider updating the rules.

**NOTE** Compatible versions of these mobile wallets have previously been listed as "GreenAddress" and "GreenBits". 

**NOTE** The GreenAddress terms of service and FAQ linked to by the client "Terms of Use" setting do not refer to the Blockstream Green wallet in any way and may be confusing to users. In particular, the 2of3 type of wallet cannot be created from the Blockstream Green wallets which makes the terms of service and FAQ confusing.

**CONCERN** By default, this wallet uses 2of2 multisig addressing such that the GreenAddress service must co-sign all transactions.  This is the same system used historically by the GreenAddress and GreenBits wallets and has been [previously reviewed](https://github.com/bitcoin-dot-org/bitcoin.org/pull/1065#issuecomment-143193918).  No additional review of that system was done here other than to verify it was the same system.  For a user to recover funds during a GreenAddress service failure, they must be in possession of pre-signed nLockTime transactions from the GreenAddress service, which requires changes to the default wallet configuration to receive these transactions.  Thus, it is suspected that there are many users that would lose access to funds if the GreenAddress service were to fail.

**NOTE** The wallet does not allow spending of unconfirmed transactions yet it lists unconfirmed transactions in the balance and does not display what the spendable balance is.  This can be confusing to users when trying to send a transaction with seemingly sufficient balance and the wallet displays "Insufficient funds."  This may be especially confusing when the unconfirmed transaction involves change from a previous send.

Basic requirements:

- Sufficient users and/or developers feedback can be found without concerning
  issues, or independent security audit(s) is available

  **NOTE** No independent security audit is known
  
  **PASS** Sufficient users and feedback found
  * Google Play: 50,000+ installs, 400 ratings
  * Reddit:  https://www.reddit.com/r/Bitcoin/comments/b3bgsj/announcing_the_allnew_blockstream_green_wallet/
  * Reddit:  https://www.reddit.com/r/Bitcoin/comments/b3059y/blockstream_launches_new_premium_bitcoin_wallet/
  * Based on previously listed GreenBits and GreenAddress
 
  
- No indication that users have been harmed considerably by any issue in
  relation to the wallet
  
  **PASS** No direct evidence of considerable harm has been found, however [GreenAddress servers do not have 100% uptime](https://www.reddit.com/r/Bitcoin/comments/3fvhh4/greenaddress_has_been_down_the_past_few_days_how/) and recovery procedures are cumbersome at best, only possible after 90 days (by default), and only possible if the user has chosen specific optional settings to receive recovery materials by email.  During past server failures, users have certainly been at the very least inconvenienced.
 
  
- No indication that security issues have been concealed, ignored, or not
  addressed correctly in order to prevent new or similar issues from happening
in the future

  **PASS** No indication found.  Responsive and willing to engage on Reddit and Bitcointalk.

- No indication that the wallet uses unstable or unsecure libraries

  **PASS** No indication found. Uses bitcoinj on Android.

- No indication that changes to the code are not properly tested

  **PASS** No indication found

- Wallet was publicly announced and released since at least 3 months

  **PASS** Eligible for listing on 19 June 2019
  
- No concerning bug is found when testing the wallet

  **PASS** No concerning bug found

- Website supports HTTPS and 301 redirects HTTP requests

  **PASS** blockstream.com supports http redirects

- SSL certificate passes [Qualys SSL Labs SSL
  test](https://www.ssllabs.com/ssltest/)
  
  **PASS** blockstream.com rating: A+
  
- Website serving executable code or requiring authentication uses HSTS with a
  max-age of at least 180 days
  
  **PASS** max-age is 365 days
  
- The identity of CEOs and/or developers is public

  **PASS** https://blockstream.com/about/
  
- Avoid address reuse by displaying a new receiving address for each transaction
  in the wallet UI
  
  **PASS** Uses a new address for each new transaction
  
- Uses deterministic ECDSA nonces (RFC 6979)

  **PASS** Uses bitcoinj
  
- Avoid address reuse by using a new change address for each transaction

  **PASS** Does not reuse change addresses
	

- User has access to private keys
- If private keys or encryption keys are stored online:

  **NOTE** The client's HD seed is stored locally encrypted by a key which is stored online (server) and accessed via a six digit PIN
  
  - Refuses weak passwords (short passwords and/or common passwords) used to
    secure access to any funds, or provides an aggressive account lock-out
feature in response to failed login attempts along with a strict account
recovery process.

     **PASS** PIN system provides aggressive lockout (3 failures, keys are erased) enforced by server.  Original HD phrase is then required for recovery.


- If user has exclusive access over its private keys:
  - Allows backup of the wallet

      **PASS** Settings allows viewing of the BIP39 seed
        
  - Restoring wallet from backup is working

      **PASS** Wallet was successfully restored using BIP39 seed.
        
  - Source code is public and kept up to date under version control system

     **PASS** https://github.com/Blockstream/green_ios

     **PASS** https://github.com/Blockstream/green_android
  
- If user has no access to some of the private keys in a multi-signature wallet:
  
  - Provides 2FA authentication feature

      **PASS** Provides a rich set of 2FA options: E-mail, Google Authenticator, SMS, and Phone call

  - Reminds the user to enable 2FA by email or in the main UI of the wallet

      **PASS** Wallet prompts the user once during wallet setup.
      
      **NOTE** Previous GreenAddress/GreenBits wallets reminded the user to set up 2FA more frequently.  This should be reconsidered since wallet recovery transactions are only emailed to users that set up email 2FA.

  - User session is not persistent, or requires authentication for spending

      **PASS** 2FA is required for spending if enabled

  - Gives control to the user over moving their funds out of the multi-signature wallet

      **PASS** Can send funds to any address or use recovery tools


- For hardware wallets:

  **N/A**

  - Uses the push model (computer malware cannot sign a transaction without user
    input)
  - Protects the seed against unsigned firmware upgrades
  - Supports importing custom seeds
  - Provides source code and/or detailed specification for blackbox testing if
    using a closed-source Secure Element

Optional criteria (some could become requirements):

- Received independent security audit(s)

  **NOTE** No known security audits

- Does not show "received from" Bitcoin addresses in the UI

  **PASS** No received from address is displayed

- Provides a bug reporting policy on the website

  **PASS** Support email, subreddit, and telegram on https://docs.blockstream.com/green/support.html

- Website serving executable code or requiring authentication is included in the
  [HSTS preload list](https://hstspreload.org/)
  
  **PASS** blockstream.com is included in the preload list
  
- If user has exclusive access over its private keys:

  **NOTE** This is a multi-sig wallet.  User has some control over private keys.

  - Supports HD wallets (BIP32)

     **PASS** BIP32 with BIP39 seed, uses m/1/i path for 2of2 multisig
  
  - Provides users with step to print or write their wallet seed on setup

     **PASS**  On setup wallet displays the BIP39 seed, requests user write it down, and quizzes a few of the words
      
  - Uses a strong KDF and key stretching for wallet storage and backups

     **PASS** Wallet storage uses strong keying material stored on server retrieved with PIN 
https://ghgreenaddress.files.wordpress.com/2014/04/greenaddressp2sh2of2hd-6.pdf
  
  - On desktop platform:

      **N/A**
  
    - Encrypt the wallet by default
- For hardware wallets:

  **N/A**
  - Prevents downgrading the firmware

crwatkins commented:
Hey @greenaddress could you review the question to you that is somewhat buried in the [above review](https://github.com/bitcoin-dot-org/bitcoin.org/pull/2901#issuecomment-507094990)?

greenaddress commented:
@crwatkins thank you for your review!

You are right in regards to iOS, we don't have support for SPV/validation so I have fixed the scoring to checkfailvalidationcentralized

I also squashed and rebased - sorry for the delay

crwatkins commented:
LGTM cd4bb50

wbnns commented:
@greenaddress 

Hello, we never heard back from you in #3023 after following up with additional questions on July 13th. We noted that you all were contacted outside of GitHub as well, but did not reply to the issue. In addition, it seems it recently took a month to reply to the last question on this thread.

You don't seem to be very accessible, and we're concerned about your lack of responsiveness combined with responsibility over a product that involves people storing their money. For example, that it might be difficult for users to get in touch with you when they have a question, since that seems to be the trouble we are having here. This concern is further elevated because by default the wallet uses a 2 of 2 multisig where the platform is required in order to co-sign all transactions. If your service were to go down, we're not sure people would be able to reach you for support, since we're having problems when there is no service interruption.

Could you please elaborate on why it has been so difficult to get in touch with you, and why this shouldn't be a concern?

Cobra-Bitcoin commented:
Maybe we should close this PR and not list this wallet? These guys take months to reply to basic questions about their wallet. Doesn't exactly inspire confidence especially when it's a custodial 2-of-2 signatures service...

crwatkins commented:
The Green family of wallets have included some of our oldest, most stable, and most feature-full wallets, however

> Doesn't exactly inspire confidence especially when it's a custodial 2-of-2 signatures service...

I have to reluctantly agree with that.  To be honest, I'm not sure that we would list a new wallet that didn't have a default recovery option.

wbnns commented:
This is scheduled to be closed on Wednesday, August 28th.

friendsofbitcoin commented:
Can Green wallet be added to Bitcoin.org now ? I originally stopped recommending it due to the forced 2of 2 but the newest version is single sig BIP 39 by default and one of the few reproducible open source non custodial android wallets. 

https://walletscrutiny.com/android/com.greenaddress.greenbits_android_wallet/

The newest versions are actively being updated from their repos, are their any other issues that need to be addressed?

crwatkins commented:
@friendsofbitcoin Thanks!  It's great to see that the 2of2 wallet is not the only choice with both a single signature wallet and a 2of3 option available.  I also see that the repos are being updated frequently and I've recently observed that Blockstream Green's support channels are extremely responsive.  I do not currently know of any other issues that need to be addressed.  As I commented above, the Green family of wallets have included some of our oldest, most stable, and most feature-full wallets, and I would like to see Blockstream Green listed.

@greenaddress Would you be willing to open a new PR to submit Green for listing?    Since this PR was was submitted, the wallet listings have changed a fair amount and there is a new **features** field that needs to be added which is described [here](https://github.com/bitcoin-dot-org/bitcoin.org/blob/master/docs/managing-wallets.md#adding-a-wallet).  Also it would be great to review the [scoring](https://github.com/bitcoin-dot-org/bitcoin.org/blob/master/docs/managing-wallets.md#score) to see if any scores might be able to be improved.

crwatkins commented:
Note that there is one issue that blockstream.com is actually on the HSTS preload list, but it no longer serves HSTS headers as it once did correctly during the last review.
